lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -5663,92 +5474,515 @@ class Differentiation : public SILModuleTransform {
+    if (auto *thunkRef = dyn_cast<FunctionRefInst>(ai->getCallee())) {
+      SILAutoDiffIndices desiredIndices(resultIndex, parameterIndices);
+      auto *thunk = thunkRef->getReferencedFunction();
+      auto newThunkName = "AD__" + thunk->getName().str() +
+          "__cloned_curry_thunk_" + desiredIndices.mangle();
+
+      auto thunkTy = thunk->getLoweredFunctionType();
+      auto thunkResult = thunkTy->getSingleResult();
+      if (auto resultFnTy = thunkResult.getType()->getAs<SILFunctionType>()) {
+        // Construct new curry thunk type with `@differentiable` result.
+        auto diffableResultFnTy = resultFnTy->getWithExtInfo(
+            resultFnTy->getExtInfo().withDifferentiable(true));
+        auto newThunkResult = thunkResult.getWithType(diffableResultFnTy);
+        auto thunkType = SILFunctionType::get(
+            thunkTy->getGenericSignature(), thunkTy->getExtInfo(),
+            thunkTy->getCoroutineKind(), thunkTy->getCalleeConvention(),
+            thunkTy->getParameters(), {}, {newThunkResult}, {},
+            thunkTy->getASTContext());
+
+        // Construct new curry think.
+        SILOptFunctionBuilder fb(transform);
+        auto *newThunk = fb.getOrCreateFunction(
+            loc, newThunkName,
+            getSpecializedLinkage(thunk, thunk->getLinkage()), thunkType,
+            thunk->isBare(), thunk->isTransparent(), thunk->isSerialized(),
+            thunk->isDynamicallyReplaceable(), ProfileCounter(),
+            thunk->isThunk());
+        if (newThunk->empty()) {
+          newThunk->setOwnershipEliminated();
+          SILFunctionCloner cloner(newThunk);
+          cloner.cloneFunction(thunk);
+        }
 
+        auto *retInst =
+            cast<ReturnInst>(newThunk->findReturnBB()->getTerminator());
         AutoDiffFunctionInst *adfi;
         {
           SILBuilder builder(retInst);
           adfi = builder.createAutoDiffFunction(loc, parameterIndices,
                                                 differentiationOrder,
                                                 retInst->getOperand());
+          resultIndices[adfi] = resultIndex;
           builder.createReturn(loc, adfi);
         }
         retInst->eraseFromParent();
 
-        newF->setOwnershipEliminated();
-
-        if (processAutoDiffFunctionInst(adfi)) {
+        getAutoDiffFunctionInsts().push_back(adfi);
+        if (processAutoDiffFunctionInst(adfi))
           return nullptr;
-        }
 
-        auto *fn = builder.createFunctionRefFor(sourceFn->getLoc(), newF);
-        SmallVector<SILValue, 8> newArgs;
-        for (auto origArg : ai->getArguments())
-          newArgs.push_back(origArg);
-
-        auto *inst =
-            builder.createApply(ai->getLoc(), fn, ai->getSubstitutionMap(),
-                                newArgs, ai->isNonThrowing());
-        return inst;
+        auto *newThunkRef = builder.createFunctionRef(loc, newThunk);
+        auto arguments = map<SmallVector<SILValue, 8>>(
```

Comment:
```suggestion
        SmallVector<SILValue, 8> arguments(ai->getArguments().begin(),
                                           ai->getArguments().end())
```

Reply:
Oh nice. I've been using the suboptimal `map<SmallVector<SILValue, ...>>` pattern in multiple places.

Reply:
Done. Also changed another `map<SmallVector<...>>` call-site to `SmallVector` direct constructor call.